### PR TITLE
Ip address retrieval done using 0.6.X api 'os.networkInterfaces()'

### PR DESCRIPTION
**Pull request by prabhakhar**
The implementation details is explained in the comment in http.request.js.

### CODE REVIEW

**Comment by s3u** [comment]
Reviewing ...

File: modules/engine/lib/engine/http.request.js [hunk 1]
```diff
@@ -705,11 +713,36 @@ function getStatus(resourceUri, statement, params, res, resource, respJson, resp
...
 }
 
-function ip() {
-    // TODO Change the implementation to return the IP Address using
-    // os.getNetworkInterfaces() call once we upgrade to node 0.5.5.
-    // For testing purposes host name is returned.
-    return os.hostname();
+function ip(cb) {
+    /*
+     * 1. Get all the non internal ip-addresses
```
Comment (s3u) [inline modules/engine/lib/engine/http.request.js:718]: Could you clarify this comment? Do you mean 'all network interfaces'?

File: modules/engine/lib/engine/http.request.js [hunk 2]
```diff
@@ -705,11 +713,36 @@ function getStatus(resourceUri, statement, params, res, resource, respJson, resp
...
+     * 2. Do a reverse lookup for all the ip-addresses got in Step 1
+     * 3. Find current machine's full hostname
+     * 4. DNS lookup hostname, got in Step 3, to get the network IP
+     *
+     * Gets Ip from the DNS server if there is no local '/etc/hosts' entry for
+     * the full hostname. See issue https://github.com/joyent/node/issues/2689
+     */
+    var hostname;
+    var ips = _.pluck(_.filter(_.flatten(_.values(os.networkInterfaces())), function (ip) {
+        return ip.internal === false && ip.family === 'IPv4';
```
Comment (s3u) [inline modules/engine/lib/engine/http.request.js:725]: What happens for IPv6 addresses?

File: modules/engine/test/request-id-test.js [hunk 3]
```diff
@@ -329,5 +329,52 @@ module.exports = {
                 test.done();
             }
         });
+    },
+    'ip-in-request-id' : function (test) {
```
Comment (s3u) [inline modules/engine/test/request-id-test.js:333]: There is too much time spent spent in serving XML, sending some HTTP requests and so on in this test. Ultimately all we care about is that you get the 127 address. Can you simplify this test?

We have some similar tests from the early days, and let's avoid such tests going forward.

File: modules/engine/lib/engine/http.request.js (hunk 2 above)
Comment (prabhakhar) [inline modules/engine/lib/engine/http.request.js:725]: My code handles only  ipv4. If we have to decide on the IP version, please let me know. I can make the corresponding changes.

File: modules/engine/test/request-id-test.js (hunk 3 above)
Comment (prabhakhar) [inline modules/engine/test/request-id-test.js:333]: Thanks. I will change/add test to do only check for the Ip address.

File: modules/engine/lib/engine/http.request.js (hunk 2 above)
Comment (s3u) [inline modules/engine/lib/engine/http.request.js:725]: I'm unclear on this - how does it differ as far as getting the ip is concerned?

Subbu

File: modules/engine/lib/engine/http.request.js (hunk 1 above)
Comment (prabhakhar) [inline modules/engine/lib/engine/http.request.js:718]: Let me clarify, For example, in one of my dev machine, if I do os.networkInterfaces(), I get

{ lo: 
   [ { address: '127.0.0.1',
       family: 'IPv4',
       internal: true },
     { address: '::1',
       family: 'IPv6',
       internal: true } ],
  'eth0:0': 
   [ { address: '10.254.35.200',
       family: 'IPv4',
       internal: false } ],
  'eth0:1': 
   [ { address: '10.254.35.201',
       family: 'IPv4',
       internal: false } ],
  'eth0:2': 
   [ { address: '10.254.35.202',
       family: 'IPv4',
       internal: false } ],
  'eth0:3': 
   [ { address: '10.254.35.203',
       family: 'IPv4',
       internal: false } ],
  'eth0:4': 
   [ { address: '10.254.35.204',
       family: 'IPv4',
       internal: false } ],
  eth0: 
   [ { address: '10.254.34.168',
       family: 'IPv4',
       internal: false },
     { address: 'fe80::be30:5bff:fec4:b6df',
       family: 'IPv6',
       internal: false } ] }

I mean, I consider all the ips returned by this call.

Comment (prabhakhar) [inline modules/engine/lib/engine/http.request.js:718]: Actually, I am cared only about the Ipv4 addresses which we have on decide upon.

File: modules/engine/test/request-id-test.js (hunk 3 above)
Comment (prabhakhar) [inline modules/engine/test/request-id-test.js:333]: As I think, ip() method does not belong to http.request.js.

Shall I create a separate module like, util.js or network-util to have this ip() method. Then, the functionality can be unit tested and of course, easy to use bu other modules.

Comment (s3u) [inline modules/engine/test/request-id-test.js:333]: I think it is okay when engine is fired without a server. But whence is running on a server, req.connection.address() gives the address.

**Comment by idralyuk** [comment]
Hmm, doing dns lookups on every request will be slow, unless we or the OS cache the result.

**Comment by prabhakhar** [comment]
@idralyuk, I didnt get you. The dns lookup is done only once (per node process). The module level networkIp variable caches it already. If it is otherwise, let me know.

**Comment by idralyuk** [comment]
Ah, I thought this was done on every request.

On Sun, Feb 5, 2012 at 4:58 PM, prabhakhar <
reply@reply.github.com

**Pull request merged**